scipy/stats/_qmc.py

```diff
@@ -481,6 +485,186 @@ def _perturb_discrepancy(sample: np.ndarray, i1: int, i2: int, k: int,
+    is_valid = np.all(np.logical_and(new_sample >= 0, new_sample <= 1), axis=1)
+    sample[is_valid] = new_sample[is_valid]
+
+    return sample
+
+
+def lloyd_centroidal_voronoi_tessellation(
+        sample: npt.ArrayLike,
+        *,
+        n_iters: IntNumber = 10,
+        seed: SeedType = None
+) -> np.ndarray:
+    """Centroidal Voronoi Tessellation.
+
+    Perturb a sample of points in :math:`[0, 1]^d` using Lloyd's algorithm.
+
+    Parameters
+    ----------
+    sample : array_like (n, d)
+        The sample to iterate on. With ``n`` the number of sample and ``d``
+        the dimension.
+    n_iters : int, optional
+        Number of iterations. Too many iterations tend to cluster the sample
+        as a hypersphere. Default is 10.
+    seed : {None, int, `numpy.random.Generator`}, optional
+        If `seed` is None the `numpy.random.Generator` singleton is used.
+        If `seed` is an int, a new ``Generator`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``Generator`` instance then that instance is
+        used.
+
+    Returns
+    -------
+    sample : array_like (n, d)
+        The sample after `n_iters` of Lloyd's algorithm.
+
+    Notes
+    -----
+    Lloyd's algorithm is an iterative process which purpose is to improve
+    the dispersion of a sample. For a given sample: (i) compute a Voronoi
+    Tessellation; (ii) find the centroid of each Voronoi cell; (iii) move the
+    samples toward the centroid of their respective cell.
+
+    The process converges to equally spaced points. It implies that measures
+    like the discrepancy could suffer from too many iterations. On the other
+    hand, L1 and L2 distances should improve. This is especially true with
+    QMC methods which tend to favor the discrepancy over other criteria.
+
+    .. warning::
+
+       The Voronoi Tessellation step is expensive and quickly becomes
+       intractable with dimensions as low as 10 even for a sample of size
+       as low as 1000.
+
+    References
+    ----------
+    .. [1] Lloyd. "Least Squares Quantization in PCM".
+       IEEE Transactions on Information Theory, 1982.
+
+    Examples
+    --------
+    >>> from scipy.spatial.distance import cdist
+    >>> from scipy.stats import qmc
+    >>> rng = np.random.default_rng()
+    >>> sample = rng.random((128, 2))
+
+    Compute the quality of the sample using the L2 criterion.
+
+    >>> def l2_norm(sample):
+    ...    l2 = cdist(sample, sample)
+    ...    return np.min(l2[l2.nonzero()])
+
+    >>> l2_norm(sample)
+    0.00547...  # random
+
+    Now process the sample using Lloyd's algorithm and check the improvement
+    on the L2. The value should increase.
+
+    >>> sample = qmc.lloyd_centroidal_voronoi_tessellation(sample)
+    >>> l2_norm(sample)
+    0.0273...  # random
+
+    """
+    sample = np.asarray(sample)
+    rng = check_random_state(seed)
+
+    if not sample.ndim == 2:
+        raise ValueError('Sample is not a 2D array')
```

Comment:
Should we reject inputs that are beyond the practical limit of Qhull (`sample.shape[1] > 9` or something)

Reply:
If there is actually such limit with the underlying library sure. Or is it just slow? The question is, should it raise or be a warning?

Reply:
Hmm, from the [Qhull site](http://www.qhull.org/)

> Qhull does not support triangulation of non-convex surfaces, mesh generation of non-convex objects, **medium-sized inputs in 9-D and higher**, alpha shapes, weighted Voronoi diagrams, Voronoi volumes, or constrained Delaunay triangulations.

Actually, it seems to take **small** 9D inputs, but I imagine drawing the line somewhere might make sense vs. allowing a segmentation fault/memory explosion/hanging process.

Reply:
@tupui I think that this can be resolved by checking what happens for large problems in 9 dimensions and higher-dimensional problems. If there is a sensible error, it could be fine without explicit checking here, or we might want to re-raise something specific to this function. If there is not an error yet, it might make sense to add the check to `Voronoi` instead of this function.